### ISSUE TITLE
[Important] Heartbeat Prompt Repeating

### ISSUE THREAD

**Issue by rayborn** [mozilla/normandy#698]
We've had a few repeated prompt reports in the past (eg [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1322425)), but unless my reading of Telemetry is very wrong (possible), this seems to be a more serious issue than I thought.

I noticed some repeated UUIDs in a non-repeated campaign, so I decided to generate a histogram of the number of dismissed prompt pings for the recipe on a per user bases.  Here's what I found:

| # Dismissals  |  # Users who dismissed N times  |  % Users who dismissed N times  |  % All Dismissals in Bucket  |
| -----  |  -----  |  -----  |  -----  |
| 1  |  8015  |  99.2%  |  96.0%  |
| 2  |  33  |  0.4%  |  0.8%  |
| 3  |  9  |  0.1%  |  0.3%  |
| 4  |  5  |  0.1%  |  0.2%  |
| 5  |  4  |  0.0%  |  0.2%  |
| 6  |  1  |  0.0%  |  0.1%  |
| 7  |  1  |  0.0%  |  0.1%  |
| 8  |  2  |  0.0%  |  0.2%  |
| 10  |  1  |  0.0%  |  0.1%  |
| 14  |  1  |  0.0%  |  0.2%  |
| 20  |  1  |  0.0%  |  0.2%  |
| 32  |  1  |  0.0%  |  0.4%  |
| 40  |  1  |  0.0%  |  0.5%  |
| 56  |  1  |  0.0%  |  0.7%  |

LMK if you have questions or would like to review the Gist.  Until investigated, this seems like a high priority issue.
I used this recipe 130 for this Spark job.

**Comment by Osmose** [comment]
@rayborn A link to the gist would be very useful, as I'd like to see what events correspond do a dismissal in that table.

**Comment by Osmose** [comment]
Rob confirmed via Slack that the issue is still happening after the merged fix.

**Comment by Osmose** [comment]
@MattGrimes @rayborn So looking through the data dump of pings, we're not seeing anything that usefully correlates affected users. We still suspect it has something to do with the reliability of client-side storage.

I anticipate that we'll release the system add-on on the release channel on Wednesday (sampled, but hopefully by the end of the week we'll be at 100%). Given that the system add-on has a different mechanism for storage, the issue might be resolved by that.

Here's my proposal:

1. Launch the system add-on and check to see if we're still getting repeat heartbeats.
2. If so, add monitoring (potentially via Telemetry?) to the add-on, so that we can see things like:
   - How often are recipes executing
   - How often are recipes throwing exceptions
   - Is client storage considered durable?
3. Once we have an implementation plan, determine if we can land monitoring with the preference-experiments update when we ship to release.

This potentially pushes out diagnosing and resolving this issue for several weeks, but having monitoring in place would be a huge improvement in our ability to diagnose issues like this. Also, given the percentage of sampled users that are seeing this, and the fact that we've been living with this for a while without knowing, I'm not too worried about fixing this _this week_.

Does that sound reasonable to ya'll?

**Comment by rayborn** [comment]
I think that this is a reasonable test plan.  It sounds like the system add-on is unlikely to have this issue, so I'd like to think that this will solve itself very shortly.  That said, I may take steps to limit volume in the interim.  I don't want to be a source of annoyance while we're aware / can avoid it.

Thanks for the work on this!

**Comment by rayborn** [comment]
This appears to be an issue post sys-addon.  I reran my query for pings after May 4th.  I've generated a ping JSON as before if it's helpful.

|Repeats|# Users in bucket|
|----|----|
|7|1|
|6|2|
|5|1|
|4|3|
|3|7|
|2|8|

**Comment by Osmose** [comment]
Next step seems to be improving Normandy monitoring, but we're going to punt on that until after opt-out study support lands thanks to shifting priorities.

**Comment by rayborn** [comment]
So I rewrote my re-prompt script to join against main summary pings (using HBase) to have a better idea of whether the repeat users that we're seeing are on the system addon or built in version.

Unfortunately, while the system addon seems better, repeats still exist.

| Repeats Present? | % Users in group with built in | % Users in group with sys addon | % Users where packet parsing fails | Total |
| ------ | ------ | ------ | ------ | ------ |
| False | 50% | 42% | 8% | 100% |
| True | 71% | 20% | 9% | 100% |

Data is for recipe Tel-Exp-2 after May 12.  Ping me for the Spark gist.  It's a bit hard to parse, but I feel pretty good about the logic.  Happy to do a code review or peer.

**Comment by mythmon** [comment]
It isn't clear that there is anything actionable here. Thanks for all the data collection @rayborn. I'm going to close this, as I don't think there is any more we can do. Cloned profiles are probably always going to be an issue, and just something we're going to have to deal with in data analysis. I'm confident that normal users aren't getting double prompted.
